### PR TITLE
Add error for Get-PSSession -ComputerName on Unix

### LINKED ISSUE #20995: Get-PSSession with -ComputerName does not work
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://learn.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

1. launch pwsh
2. execute command: Get-PSSession -ComputerName dc01

---
_This has already been stated in #6321 (and #17768?) but those were closed due to inactivity and as the bot closed those for housekeeping, I now create a new one._

### Expected behavior

```console
Id    Name    ComputerName    ComputerType    State    ConfigurationName    Availability
--    ----    ------------    ------------    -----    -----------------    ------------
1     WinRM1  dc01            RemoteMachine   Opened   Microsoft.PowerShell Available



_(copied from a Windows machine executing the same command)_
```


### Actual behavior

```console
PS> Get-PSSession -ComputerName dc01       
                                                                                                                                                                                                                                       
Get-PSSession: The term 'New-WSManSessionOption' is not recognized as a name of a cmdlet, function, script file, or executable program.
Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
```


### Error details

_No response_

### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.4.0
PSEdition                      Core
GitCommitId                    7.4.0
OS                             Kali GNU/Linux Rolling
Platform                       Unix
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```


### Visuals

_No response_

### PR DESCRIPTION
**Pull request by jborean93**
# PR Summary
Emit an exception when attempting to use Get-PSSession -ComputerName on a non-Windows platform that makes it clear it cannot be used on those platforms. Now when running this on non-Windows it will emit the following error.

```powershell
PS /home/jborean/dev/PowerShell> Get-PSSession -ComputerName foo
Get-PSSession: Remote PSSession enumeration with -ComputerName is only supported on Windows and not "Arch Linux".
```

## PR Context
I don't think this feature will ever be implemented, even if it is it would be nice to error with a proper message not the current

> Get-PSSession: The term 'New-WSManSessionOption' is not recognized as a name of a cmdlet, function, script file, or executable program.
> Check the spelling of the name, or if a path was included, verify that the path is correct and try again.

Fixes: https://github.com/PowerShell/PowerShell/issues/20995

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

**Comment by iSazonov** [comment]
There was WG conclusion about root of the issue https://github.com/PowerShell/PowerShell/issues/6321#issuecomment-1092132470

**Comment by jborean93** [comment]
> There was WG conclusion about root of the issue https://github.com/PowerShell/PowerShell/issues/6321#issuecomment-1092132470

I think Paul may have been confused there, `Get-PSSession -ComputerName` is and has always been focused purely on WSMan. It sends a SOAP request to the WSMan service to ask it to enumerate the number of shells open and has it return that list. There is nothing here that can be replicated under ssh and even if there was it would be part of a new `-HostName` parameter to match up with how other PSRemoting cmdlets like `Invoke-Command` and `Enter-PSSession`. In short getting `-ComputerName` to work on non-Windows requires the PowerShell team to start shipping the `Microsoft.WSMan.Management` module on non-Windows which requires more work with the omi library to get working properly. They've repeatedly said in the past they have don't want anything to do with WSMan anymore so instead of waiting forever for this to be re-implemented I thought at least a better error saying this only works on Windows is best.

**Comment by iSazonov** [comment]
Hmm, I see Paul said that we can enumerate `PSSession` objects which was created in current process.

**Comment by jborean93** [comment]
Yep and using -ComputerName is for sessions on a completely different host. In fact it uses a mechanism in the WSMan service and not PowerShell itself to track the sessions on the remote host. Maybe how PowerShell tracks that information can be expanded in the future but it shouldn’t hold up something like this.

**Comment by iSazonov** [comment]
@jborean93 For my education, does ` -ComputerName` return all remote pssessions on the remote computer or only sessions created from the current pwsh process?

**Comment by jborean93** [comment]
When you do `Get-PSSession -ComputerName foo` PowerShell internally runs the following:

```powershell
$getParams = @{
    ResourceURI = 'Shell'
    Enumerate = $true
    ComputerName = 'foo'
    SessionOption = (New-WSManSessionOption)
}
Get-WSManInstance @getParams
```

In protocol terms this sends the following WSMan POST message to `foo`:

```xml
<s:Envelope xmlns:s="http://www.w3.org/2003/05/soap-envelope" xmlns:a="http://schemas.xmlsoap.org/ws/2004/08/addressing" xmlns:n="http://schemas.xmlsoap.org/ws/2004/09/enumeration" xmlns:w="http://schemas.dmtf.org/wbem/wsman/1/wsman.xsd" xmlns:p="http://schemas.microsoft.com/wbem/wsman/1/wsman.xsd" xmlns:b="http://schemas.dmtf.org/wbem/wsman/1/cimbinding.xsd">
  <s:Header>
    <a:To>http://foo:5985/wsman</a:To>
    <w:ResourceURI s:mustUnderstand="true">http://schemas.microsoft.com/wbem/wsman/1/windows/shell</w:ResourceURI>
    <a:ReplyTo>
      <a:Address s:mustUnderstand="true">http://schemas.xmlsoap.org/ws/2004/08/addressing/role/anonymous</a:Address>
    </a:ReplyTo>
    <a:Action s:mustUnderstand="true">http://schemas.xmlsoap.org/ws/2004/09/enumeration/Enumerate</a:Action>
    <w:MaxEnvelopeSize s:mustUnderstand="true">512000</w:MaxEnvelopeSize>
    <a:MessageID>uuid:B03C7765-66AC-4DAB-B84D-BC62464A1AC8</a:MessageID>
    <w:Locale xml:lang="en-US" s:mustUnderstand="false"/>
    <p:DataLocale xml:lang="en-US" s:mustUnderstand="false"/>
    <p:SessionId s:mustUnderstand="false">uuid:00DC848B-B247-485C-B144-EB709AA1586A</p:SessionId>
    <p:OperationID s:mustUnderstand="false">uuid:CB4C0368-D3A8-45D0-AE6E-2C1B90FAD70E</p:OperationID>
    <p:SequenceId s:mustUnderstand="false">1</p:SequenceId>
    <w:OperationTimeout>PT60.000S</w:OperationTimeout>
  </s:Header>
  <s:Body>
    <n:Enumerate>
      <w:OptimizeEnumeration/>
      <w:MaxElements>32000</w:MaxElements>
    </n:Enumerate>
  </s:Body>
</s:Envelope>
```

This `Enumerate` action tells the WSMan service on `foo` to return all "shells" that have been opened on the service. Note there is no client information like the process or host it is from which means `foo` will be returning all shells it knows about. After receiving this request `foo` will return the following `EnumerateResponse`

```xml
<s:Envelope xmlns:s="http://www.w3.org/2003/05/soap-envelope" xmlns:a="http://schemas.xmlsoap.org/ws/2004/08/addressing" xmlns:n="http://schemas.xmlsoap.org/ws/2004/09/enumeration" xmlns:w="http://schemas.dmtf.org/wbem/wsman/1/wsman.xsd" xmlns:p="http://schemas.microsoft.com/wbem/wsman/1/wsman.xsd" xml:lang="en-US">
  <s:Header>
    <a:Action>http://schemas.xmlsoap.org/ws/2004/09/enumeration/EnumerateResponse</a:Action>
    <a:MessageID>uuid:142D87D6-7CBC-4E9E-9D4F-95CD0EE5CAA2</a:MessageID>
    <a:To>http://schemas.xmlsoap.org/ws/2004/08/addressing/role/anonymous</a:To>
    <p:OperationID s:mustUnderstand="false">uuid:CB4C0368-D3A8-45D0-AE6E-2C1B90FAD70E</p:OperationID>
    <p:SequenceId>1</p:SequenceId>
    <a:RelatesTo>uuid:B03C7765-66AC-4DAB-B84D-BC62464A1AC8</a:RelatesTo>
  </s:Header>
  <s:Body>
    <n:EnumerateResponse>
      <n:EnumerationContext/>
      <w:Items>
        <rsp:Shell xmlns:rsp="http://schemas.microsoft.com/wbem/wsman/1/windows/shell">
          <rsp:ShellId>E578F8FB-5A82-4491-AC46-E04EDBD41CD7</rsp:ShellId>
          <rsp:Name>WinRM1</rsp:Name>
          <rsp:ResourceUri>http://schemas.microsoft.com/powershell/Microsoft.PowerShell</rsp:ResourceUri>
          <rsp:Owner>DOMAIN\vagrant-domain</rsp:Owner>
          <rsp:ClientIP>192.168.56.12</rsp:ClientIP>
          <rsp:ProcessId>832</rsp:ProcessId>
          <rsp:IdleTimeOut>PT7200.000S</rsp:IdleTimeOut>
          <rsp:InputStreams>stdin pr</rsp:InputStreams>
          <rsp:OutputStreams>stdout</rsp:OutputStreams>
          <rsp:MaxIdleTimeOut>PT2147483.647S</rsp:MaxIdleTimeOut>
          <rsp:Locale>en-US</rsp:Locale>
          <rsp:DataLocale>en-US</rsp:DataLocale>
          <rsp:CompressionMode>XpressCompression</rsp:CompressionMode>
          <rsp:ProfileLoaded>Yes</rsp:ProfileLoaded>
          <rsp:Encoding>UTF8</rsp:Encoding>
          <rsp:BufferMode>Block</rsp:BufferMode>
          <rsp:State>Connected</rsp:State>
          <rsp:ShellRunTime>P0DT0H2M29S</rsp:ShellRunTime>
          <rsp:ShellInactivity>P0DT0H2M29S</rsp:ShellInactivity>
          <rsp:MemoryUsed>62MB</rsp:MemoryUsed>
          <rsp:ChildProcesses>0</rsp:ChildProcesses>
        </rsp:Shell>
        <rsp:Shell xmlns:rsp="http://schemas.microsoft.com/wbem/wsman/1/windows/shell">
          <rsp:ShellId>D76BA8D1-C52A-4E10-8A28-46262426AC09</rsp:ShellId>
          <rsp:Name>Runspace1</rsp:Name>
          <rsp:ResourceUri>http://schemas.microsoft.com/powershell/PowerShell.7</rsp:ResourceUri>
          <rsp:Owner>DOMAIN\vagrant-domain</rsp:Owner>
          <rsp:ClientIP>192.168.56.12</rsp:ClientIP>
          <rsp:ProcessId>6900</rsp:ProcessId>
          <rsp:IdleTimeOut>PT7200.000S</rsp:IdleTimeOut>
          <rsp:InputStreams>stdin pr</rsp:InputStreams>
          <rsp:OutputStreams>stdout</rsp:OutputStreams>
          <rsp:MaxIdleTimeOut>PT2147483.647S</rsp:MaxIdleTimeOut>
          <rsp:Locale>en-US</rsp:Locale>
          <rsp:DataLocale>en-US</rsp:DataLocale>
          <rsp:CompressionMode>XpressCompression</rsp:CompressionMode>
          <rsp:ProfileLoaded>Yes</rsp:ProfileLoaded>
          <rsp:Encoding>UTF8</rsp:Encoding>
          <rsp:BufferMode>Block</rsp:BufferMode>
          <rsp:State>Connected</rsp:State>
          <rsp:ShellRunTime>P0DT0H2M18S</rsp:ShellRunTime>
          <rsp:ShellInactivity>P0DT0H2M16S</rsp:ShellInactivity>
          <rsp:MemoryUsed>97MB</rsp:MemoryUsed>
          <rsp:ChildProcesses>1</rsp:ChildProcesses>
        </rsp:Shell>
      </w:Items>
      <w:EndOfSequence/>
    </n:EnumerateResponse>
  </s:Body>
</s:Envelope>
```

In this case `foo` has 2 shells that are open, one for `Microsoft.PowerShell` and another for `PowerShell.7`. They are both `Connected` which means they are still active and in use by a client. These could be `Disconnected` and waiting to be reconnected but the point is that the WSMan service reports all shells it has open. PowerShell then uses the information of each shell and filters out the results that don't match whatever the filter criteria specified by the user was.

Ultimately it's all sessions managed by the remote service not just ones opened by the current process/computer. This is why `-ComputerName` is purely a WSMan thing as it has no idea about SSH PSSessions or any other connection type.

**Comment by iSazonov** [comment]
Thanks for the full answer!

I believe the user must have the appropriate permissions for WSMan to return the list of sessions. In this case, the user could connect to the remote computer via SSH, get a list of all pwsh processes and even connect to them via local socket. Maybe this is what Paul meant.

**Comment by jborean93** [comment]
> I believe the user must have the appropriate permissions for WSMan to return the list of sessions

It certainly needs permission to perform the `Enumerate` action but I'm not aware of any extra permissions applied to what WSMan returns. Typically the access checks is done when you attempt to open one of the PSSessions as now you need to reauthenticate as the same user it was created with.

> In this case, the user could connect to the remote computer via SSH, get a list of all pwsh processes and even connect to them via local socket. Maybe this is what Paul meant.

Potentially, honestly I don't see `Get-PSSession` continuing to be used for enumerating PSSession's remotely. It could be expanded in the future to somehow track active PSSessions on the current host using a common mechanism in PowerShell and then combined with `Invoke-Command` to provide remote access. The benefit of `Get-PSSession -ComputerName ...` on Windows is the ability to get disconnected sessions and re-connect them on the new client. As WSMan is the only PSRemoting protocol that supports disconnected sessions (and custom transports cannot do so) then it makes sense only keeping this as Windows only.

**Comment by iSazonov** [comment]
> As WSMan is the only PSRemoting protocol that supports disconnected sessions (and custom transports cannot do so)

Eh, this will definitely be in demand for the management tool that is PowerShell. If WSMan does this, then users will expect the same from any custom transport that replaces WSMan. It would be better for UX if the base cmdlets didn't change and were independent of transport.

**Comment by jborean93** [comment]
I honestly feel like PSRemoting is inferior in pretty much most circumstances for interactive use which is where you typically want disconnection support. Using pure ssh with tools like tmux offer such a better experience and will be what I typically recommend going forward.

Someone may want to reimplement disconnection support for custom transports but it’ll require a lot of work and an even more special server side protocol to make use of it. Until that time I think it’s safe to just make this functionality Windows only.

**Comment by jborean93** [comment]
@SteveL-MSFT it's been over 3 months without any review by someone from the PowerShell team. Any chance this can be added to the next cycle.

File: src/System.Management.Automation/resources/RemotingErrorIdStrings.resx
Comment (iSazonov) [inline src/System.Management.Automation/resources/RemotingErrorIdStrings.resx:127]: Makes no sense to point OS version. Please remove:
```suggestion
    <value>Remote PSSession enumeration with -ComputerName is only supported on Windows.</value>
```

File: src/System.Management.Automation/engine/remoting/commands/getrunspacecommand.cs
Comment (iSazonov) [inline src/System.Management.Automation/engine/remoting/commands/getrunspacecommand.cs:346]: Our common pattern is to check whether a parameter presents.
```c#
if (MyInvocation.BoundParameters.ContainsKey(nameof(ComputerName)))
```

**Review by iSazonov — CHANGES_REQUESTED** [review]
@jborean93 Please add one test for Unix. 
I believe we can merge (after you address my requests) since it is obvious improvement.

File: src/System.Management.Automation/resources/RemotingErrorIdStrings.resx
Comment (jborean93) [reply src/System.Management.Automation/resources/RemotingErrorIdStrings.resx:127]: I find it provides a few extra debug details and doesn't hurt to include. It's also somewhat helpful to identify where the cmdlet was run when it comes to remoting ones as people may be confused if the OS platform is local or remote based. Including the local OS information in the error helps to avoid that ambiguity.

File: src/System.Management.Automation/engine/remoting/commands/getrunspacecommand.cs
Comment (jborean93) [reply src/System.Management.Automation/engine/remoting/commands/getrunspacecommand.cs:346]: Is there a particular reason for this. The code uses the `ComputerName` property value and thus it makes more sense to do the check on the actual property rather than a pwsh-ism here. It's also cleaner and shorter than accessing a dictionary to check if a key exists vs just checking the property value.

**Review by SeeminglyScience — APPROVED** [review]
LGTM, thanks Jordan!

**Review by iSazonov — APPROVED** [review]

**Pull request merged**
